Sources/StreamChatUI/MessageActionsPopup/ChatMessageActionsVC.swift

```diff
@@ -97,40 +104,62 @@ open class ChatMessageActionsVC: _ViewController, ThemeProvider {
+            let canReceiveReadEvents = channel?.ownCapabilities.contains(.readEvents) ?? channelConfig.readEventsEnabled
+            let canUpdateAnyMessage = channel?.ownCapabilities.contains(.updateAnyMessage) ?? false
+            let canUpdateOwnMessage = channel?.ownCapabilities.contains(.updateOwnMessage) ?? true
+            let canDeleteAnyMessage = channel?.ownCapabilities.contains(.deleteAnyMessage) ?? false
+            let canDeleteOwnMessage = channel?.ownCapabilities.contains(.deleteOwnMessage) ?? true
+
+            if canQuoteMessage {
                 actions.append(inlineReplyActionItem())
             }
 
-            if channelConfig.repliesEnabled && !message.isPartOfThread {
+            if canSendReply && !message.isPartOfThread {
                 actions.append(threadReplyActionItem())
             }
 
-            if channelConfig.readEventsEnabled && (!message.isPartOfThread || message.showReplyInChannel) {
+            if canReceiveReadEvents && (!message.isPartOfThread || message.showReplyInChannel) {
                 actions.append(markUnreadActionItem())
             }
 
             actions.append(copyActionItem())
 
-            if message.isSentByCurrentUser {
-                actions += [editActionItem(), deleteActionItem()]
+            if canUpdateAnyMessage {
+                actions.append(editActionItem())
+            } else if canUpdateOwnMessage && message.isSentByCurrentUser {
+                actions.append(editActionItem())
+            }
 
-            } else {
-                actions += [flagActionItem()]
+            if canDeleteAnyMessage {
+                actions.append(deleteActionItem())
+            } else if canDeleteOwnMessage && message.isSentByCurrentUser {
+                actions.append(deleteActionItem())
+            }
 
-                if channelConfig.mutesEnabled {
-                    let isMuted = currentUser.mutedUsers.contains(message.author)
-                    actions.append(isMuted ? unmuteActionItem() : muteActionItem())
-                }
+            if !message.isSentByCurrentUser {
+                actions.append(flagActionItem())
+            }
+
+            if channelConfig.mutesEnabled && !message.isSentByCurrentUser {
+                let isMuted = currentUser.mutedUsers.map(\.id).contains(message.author.id)
+                actions.append(isMuted ? unmuteActionItem() : muteActionItem())
             }
 
             return actions
-        case .pendingSend, .sendingFailed, .pendingSync, .syncingFailed, .deletingFailed:
+        case .sendingFailed:
+            return [
+                resendActionItem(),
+                editActionItem(),
+                deleteActionItem()
+            ]
+        case .pendingSend, .pendingSync, .syncingFailed, .deletingFailed:
             return [
-                (message.localState == .sendingFailed || message.failedToBeSentDueToModeration) ? resendActionItem() : nil,
```

Comment:
what would the impact to moderation be with this?

Reply:
No impact at all, this code was very redundant. It was basically doing this:
```message.locaState == .sendingFailed || (message.sendingFailed && message.isBounced == true```

So, it is the samething as just checking if the message is sendingFailed or not